docs/contributor_guide.rst

```diff
@@ -1,66 +1,110 @@
+- All commits in a PR will always be **squashed and merged into master as a single commit**, to keep a development history, make sure you break down your implementation into smaller PRs.
 - If you are making multiple PRs
 
- - Independent PRs should be based on **different** branches forking from ``master``;
- - PRs with dependencies should be raised only after all prerequisite PRs are merged into ``master``.
+  - Independent PRs should be based on **different** branches forking from ``master``;
+  - PRs with dependencies should be raised only after all prerequisite PRs are merged into ``master``.
 
 - All PRs should ideally come with corresponding **tests**;
 - All PRs should come with **documentation update**, except for internal compiler implementations;
-- All PRs should always be rebased onto the HEAD of master before merging;
-- All PRs should pass continuous integration tests (build + testing for Mac/Windows) before they get merged;
-- **PR authors should not squash commits**. Whether squashing all commits in a PR or not, will be decided by the reviewer who merges the PR into the master branch, depending on how trivially correct the commits are.
+- All PRs should always be **rebased** onto the latest master branch before merging;
+- All PRs should pass **continuous integration tests** before they get merged;
+- PR authors **should not squash commits on their own**;
+- PR titles should follow :ref:`prtag`;
+- If you want early feedback from core developers, open a PR in **Draft** state on GitHub so that you can share your progress.
 
 
 Reviewing & PR merging
 ----------------------
 
-- Error-prone commits such as IR passes & codegen will be rebased on master (**without squashing**) once approved;
-- Other commits with more trivial correctness (e.g. examples, GUI, benchmark cases, typo fixes) will first get squashed into a single commit and then rebased on master, for a cleaner master commit log.
+- The merger should always **squash and merge** PRs into the master branch;
+- The master branch is required to have a **linear history**;
+- Make sure the title follows :ref:`prtag`.
 
 
-Tips on Taichi compiler development
---------------------------------------------------
+.. _prtag:
+
+PR title tags
+-------------
+Please always prepend exactly one tag such as ``[Metal]`` to PR titles. For example, "[Metal] Support bitmasked SNode", "[OpenGL] AtomicMin/Max support", or "[Opt] Enhanced constant folding".
+
+Existing tags:
+
+- ``[Metal], [OpenGL], [CPU], [CUDA], [AMDGPU]``: backends;
+- ``[Lang]``: frontend language features, including syntax sugars;
+- ``[Std]``: standard library, e.g. `ti.Matrix` and `ti.Vector`;
+- ``[IR]``: intermediate representation;
+- ``[Opt]``: IR optimization passes;
+- ``[Type]``: type system;
+- ``[Infra]``: general infrastructure, e.g. logging, image reader;
+- ``[GUI]``: the built-in GUI system;
+- ``[CLI]``: commandline interfaces, e.g. the ``ti`` command;
```

Comment:
Is that possible to make issue titles also follow this rule?

Reply:
Currently, I don't see an immediate need for this, but if someday in the future there turns out such a need for issues, then sure :-)